File: 2183-count-array-pairs-divisible-by-k/2183-count-array-pairs-divisible-by-k.py
from math import gcd
class Solution:
    def countPairs(self, nums: List[int], k: int) -> int:
        ump = {}
        ans = 0
        for i in range(len(nums)):
            currgcd = gcd(nums[i],k)
            for x in ump:
                x1 = int(x)
                if ( x1* currgcd) %k==0:
                    ans+= ump[x]
            if currgcd in ump:
                ump[currgcd]+=1
            else:
                ump[currgcd] =1
        return ans

File: 2183-count-array-pairs-divisible-by-k/2183-count-array-pairs-divisible-by-k.cpp
class Solution {
public:
    long long countPairs(vector<int>& nums, int k) {
        unordered_map<long long,long long>ump;
        long long ans =0;
        for (int i =0;i<nums.size();++i){
            long long currgcd = __gcd(k,nums[i]);
            for(auto &[x,y]:ump){
                if((currgcd*x)%k==0){
                    ans +=y;
                }
            }
            ++ump[currgcd];
        }
        return ans;
    }
};